[CSHARP] Conflict with model definition named DateTime

<!--
Please follow the issue template below for bug reports and feature requests.
Also please indicate in the issue title which language/library is concerned. Eg:  [JAVA] Bug generating foo with bar 
-->

##### Description

Having an API with `DateTime` named definition creates conflict with csharp output ocde after generation

##### Swagger-codegen version

2.4.x, 3.0.19

##### Swagger declaration file content or url
```yaml
  Order:
    type: "object"
    properties:
      id:
        type: "integer"
        format: "int64"
      petId:
        type: "integer"
        format: "int64"
      quantity:
        type: "integer"
        format: "int32"
      shipDate:
        $ref: "#/definitions/DateTime"
      status:
        type: "string"
        description: "Order Status"
        enum:
        - "placed"
        - "approved"
        - "delivered"
      complete:
        type: "boolean"
        default: false
    xml:
      name: "Order"
  
  DateTime:
    type: string    
```

Output code generates error at build running `build.sh` file:
```console
src/IO.Swagger/Model/Order.cs(115,16): error CS0453: The type `IO.Swagger.Model.DateTime' must be a non-nullable value type in order to use it as type parameter `T' in the generic type or method `System.Nullable<T>'
src/IO.Swagger/Model/Order.cs(188,26): error CS0019: Operator `==' cannot be applied to operands of type `IO.Swagger.Model.DateTime?' and `IO.Swagger.Model.DateTime?'


```
##### Suggest a fix/enhancement
The problem is happening because a DateTime definition `string` type is created as a class. A solution for this could be stop generating models for primitive types and use their types instead the class named generated.

--- a/modules/swagger-codegen-cli/src/main/java/io/swagger/codegen/cmd/Generate.java
+++ b/modules/swagger-codegen-cli/src/main/java/io/swagger/codegen/cmd/Generate.java
@@ -280,7 +280,7 @@ public class Generate implements Runnable {
         }
 
         if (skipAliasGeneration != null) {
-            configurator.setSkipAliasGeneration(removeOperationIdPrefix);
+            configurator.setSkipAliasGeneration(skipAliasGeneration);
         }
 
         if (ignoreImportMappings != null) {

--- a/modules/swagger-codegen/src/main/java/io/swagger/codegen/DefaultCodegen.java
+++ b/modules/swagger-codegen/src/main/java/io/swagger/codegen/DefaultCodegen.java
@@ -3995,4 +3995,14 @@ public class DefaultCodegen {
     public boolean defaultIgnoreImportMappingOption() {
         return false;
     }
+
+    protected boolean isModelObject(ModelImpl model) {
+        if ("object".equalsIgnoreCase(model.getType())) {
+            return true;
+        }
+        if ((StringUtils.EMPTY.equalsIgnoreCase(model.getType()) || model.getType() == null) && (model.getProperties() != null && !model.getProperties().isEmpty())) {
+            return true;
+        }
+        return false;
+    }
 }

--- a/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/CSharpClientCodegen.java
+++ b/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/CSharpClientCodegen.java
@@ -491,7 +491,7 @@ public class CSharpClientCodegen extends AbstractCSharpCodegen {
     @Override
     public String getSwaggerType(Property p) {
         String swaggerType = super.getSwaggerType(p);
-        if (p instanceof RefProperty) {
+        if (p instanceof RefProperty && this.swagger != null) {
             final Map<String, Model> allDefinitions = this.swagger.getDefinitions();
             final Model referencedModel = allDefinitions.get(swaggerType);
             if (referencedModel == null) {

--- a/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/CSharpClientCodegen.java
+++ b/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/CSharpClientCodegen.java
@@ -4,8 +4,13 @@ import com.google.common.collect.ImmutableMap;
 import com.samskivert.mustache.Mustache;
 import io.swagger.codegen.*;
 import io.swagger.models.Model;
+import io.swagger.models.ModelImpl;
+import io.swagger.models.Swagger;
 import io.swagger.models.properties.ArrayProperty;
 import io.swagger.models.properties.Property;
+import io.swagger.models.properties.PropertyBuilder;
+import io.swagger.models.properties.RefProperty;
+import org.apache.commons.lang3.StringUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -28,6 +33,8 @@ public class CSharpClientCodegen extends AbstractCSharpCodegen {
 
     // Defines the sdk option for targeted frameworks, which differs from targetFramework and targetFrameworkNuget
     private static final String MCS_NET_VERSION_KEY = "x-mcs-sdk";
+
+    protected Swagger swagger;
 
     protected String packageGuid = "{" + java.util.UUID.randomUUID().toString().toUpperCase() + "}";
     protected String clientPackage = "IO.Swagger.Client";
@@ -433,6 +440,10 @@ public class CSharpClientCodegen extends AbstractCSharpCodegen {
 
         additionalProperties.put("apiDocPath", apiDocPath);
         additionalProperties.put("modelDocPath", modelDocPath);
+
+        if (skipAliasGeneration == null) {
+            skipAliasGeneration = true;
+        }
     }
 
     public void setModelPropertyNaming(String naming) {
@@ -475,6 +486,26 @@ public class CSharpClientCodegen extends AbstractCSharpCodegen {
         }
 
         return objs;
+    }
+
+    @Override
+    public String getSwaggerType(Property p) {
+        String swaggerType = super.getSwaggerType(p);
+        if (p instanceof RefProperty) {
+            final Map<String, Model> allDefinitions = this.swagger.getDefinitions();
+            final Model referencedModel = allDefinitions.get(swaggerType);
+            if (referencedModel == null) {
+                return swaggerType;
+            }
+            if (referencedModel instanceof ModelImpl) {
+                final ModelImpl model = (ModelImpl) referencedModel;
+                if (!this.isModelObject(model) && (model.getEnum() == null || model.getEnum().isEmpty())) {
+                    final Property property = PropertyBuilder.build(model.getType(), model.getFormat(), null);
+                    swaggerType = getSwaggerType(property);
+                }
+            }
+        }
+        return swaggerType;
     }
 
     @Override
@@ -802,5 +833,10 @@ public class CSharpClientCodegen extends AbstractCSharpCodegen {
     public Mustache.Compiler processCompiler(Mustache.Compiler compiler) {
         // To avoid unexpected behaviors when options are passed programmatically such as { "supportsAsync": "" }
         return super.processCompiler(compiler).emptyStringIsFalse(true);
+    }
+
+    @Override
+    public void preprocessSwagger(Swagger swagger) {
+        this.swagger = swagger;
     }
 }

